File: Problem038-Count and Say/Solution.py
"""
 @SEU(master);@ZZULI(bachelor)
 author:CuiXuan
 email:873059043@qq.com
 If you find any errors, please let me know.
 If you have any better solution, please let me know.

 Solution Using time 64ms
"""

class Solution:
    # @param {integer} n
    # @return {string}
    def countAndSay(self, n):
        seq=['1'];top=1;
        while n-1>0:
            n-=1;index=0;bak=[]
            i=0
            while i<top:
                num=1
                while i+1<top and seq[i+1]==seq[i]:i+=1;num+=1
                bak.append(chr(num+ord('0')))
                bak.append(seq[i])
                i+=1
            seq=bak;top=len(bak)
        return ''.join(seq)

File: Problem038-Count and Say/Solution.java
/*
 * @SEU(master);@ZZULI(bachelor)
 * author:CuiXuan
 * email:873059043@qq.com
 * If you find any errors, please let me know.
 * If you have any better solution, please let me know.
 * 
 * Solution Using time 208ms
 */
public class Solution {
    public String countAndSay(int n) {
        char[] seq=new char[100000];
        char[] bak=new char[100000];
        char[] tmp;
        char t;
        int top=1,index,l,r,num;
        seq[0]='1';seq[1]=0;
        while(--n >0){
            index=0;
            for(int i=0;i<top;i++){
                num=1;
                while(i+1<top && seq[i+1]==seq[i]){i++;num++;}
                l=index;
                while(num>0){
                    bak[index++]=(char)(num%10+'0');
                    num/=10;
                }
                r=index-1;
                while(l<r){t=bak[l];bak[l]=bak[r];bak[r]=t;l++;r--;}
                bak[index++]=seq[i];
            }
            top=index;
            tmp=seq;seq=bak;bak=tmp;
        }
        return new String(seq,0,top);
    }
}